Quantum-Control-Applications/ESR/T2.py

```diff
@@ -0,0 +1,172 @@
+            reset_phase("resonator")
+            reset_frame("ensemble")
+
+            echo = declare_stream(adc_trace=True)
+
+            assign(pulse_delay, delay_len - pi_half_len * 0.5 - pi_len * 0.5)
+            assign(readout_delay, delay_len - pi_len * 0.5 - readout_len * 0.125 - 5)
+
+            wait(8, "switch_1", "switch_2")
+            play("activate", "switch_1", duration=pi_half_len)
+            play("activate", "switch_2", duration=pi_half_len)
+            play("const", "ensemble", duration=pi_half_len)
+
+            wait(pulse_delay, "ensemble", "switch_1", "switch_2")
+
+            frame_rotation_2pi(-0.5, "ensemble")
+            play("activate", "switch_1", duration=pi_len)
+            play("activate", "switch_2", duration=pi_len)
+            play("const", "ensemble", duration=pi_len)
+
+            align()  # global align
+
+            wait(readout_delay, "resonator", "switch_receiver")
+
+            play("activate_resonator", "switch_receiver")
+            measure(
+                "readout",
+                "resonator",
+                echo,
+                dual_demod.full("cos", "out1", "sin", "out2", I),
+                dual_demod.full("minus_sin", "out1", "cos", "out2", Q),
+            )
+            save(I, I_st)
+            save(Q, Q_st)
+        save(n, n_st)
+
+    with stream_processing():
+        echo.input1().buffer(len(delays)).average().save("echo1")
+        echo.input2().buffer(len(delays)).average().save("echo2")
+        I_st.buffer(len(delays)).average().save("I")
+        Q_st.buffer(len(delays)).average().save("Q")
+        n_st.save("iteration")
+
+################################
+# Open quantum machine manager #
+################################
+
+qmm = QuantumMachinesManager()
+
+#######################
+# Simulate or execute #
+#######################
+
+simulate = True
+
+if simulate:
+    # simulation properties
+    simulate_config = SimulationConfig(
+        duration=2000,
+        include_analog_waveforms=True,
+        simulation_interface=LoopbackInterface(([("con1", 3, "con1", 1), ("con1", 4, "con1", 2)]), latency=180),
+    )
+    job = qmm.simulate(config, T2, simulate_config)  # do simulation with qmm
+    job.get_simulated_samples().con1.plot()  # visualize played pulses
+
+    analog_wf = job.simulated_analog_waveforms()
+
+    ver_t1 = (
+        analog_wf["elements"]["ensemble"][2]["timestamp"] + (analog_wf["elements"]["ensemble"][2]["duration"] / 2)
+    ) - (analog_wf["elements"]["ensemble"][0]["timestamp"] + (analog_wf["elements"]["ensemble"][0]["duration"] / 2))
+    ver_t3 = (
+        analog_wf["elements"]["resonator"][0]["timestamp"] + (analog_wf["elements"]["resonator"][0]["duration"] / 2)
+    ) - (analog_wf["elements"]["ensemble"][2]["timestamp"] + (analog_wf["elements"]["ensemble"][2]["duration"] / 2))
+
+    print("center to center time between 1st and 2nd pulse", ver_t1)
+    print("center to center time between readout and 2nd pulse", ver_t3)
```

Comment:
Is this done before executing on hardware to get the readout and pulse delays?, If so I would write in a comment

Reply:
hes